README
NAME
    Apache::Gallery - mod_perl handler to create an image gallery

SYNOPSIS
    See the INSTALL file in the distribution for installation instructions.

DESCRIPTION
    Apache::Gallery creates an thumbnail index of each directory and allows
    viewing pictures in different resolutions. Pictures are resized on the
    fly and cached. The gallery can be configured and customized in many
    ways and a custom copyright image can be added to all the images without
    modifying the original.

    The gallery will also show videos and allow downloading of other files.

CONFIGURATION
    In your httpd.conf you set the global options for the gallery. You can
    also override each of the options in .htaccess files in your gallery
    directories.

    The options are set in the httpd.conf/.htaccess file using the syntax:
    PerlSetVar OptionName 'value'

    Example: PerlSetVar GalleryCacheDir '/var/cache/www/'

    GalleryCacheDir
        Directory where Apache::Gallery should create its cache with scaled
        pictures. The default is /var/cache/www/ . Here, a directory for
        each virtualhost or location will be created automatically. Make
        sure your webserver has write access to the CacheDir.

    GalleryTemplateDir
        Full path to the directory where you placed the templates. This
        option can be used both in your global configuration and in
        .htaccess files, this way you can have different layouts in
        different parts of your gallery.

        No default value, this option is required.

    GalleryInfo
        With this option you can define which EXIF information you would
        like to present from the image. The format is: '<MyName =>
        KeyInEXIF, MyOtherName => OtherKeyInEXIF'

        Examples of keys: ShutterSpeedValue, ApertureValue, SubjectDistance,
        and Camera

        You can view all the keys from the EXIF header using this
        perl-oneliner:

        perl "-e" 'use Data::Dumper; use Image::Info qw(image_info); print
        Dumper(image_info(shift));' filename.jpg

        Default is: 'Picture Taken => DateTimeOriginal, Flash => Flash'

    GallerySizes
        Defines which widths images can be scaled to. Images cannot be
        scaled to other widths than the ones you define with this option.

        The default is '640 800 1024 1600'

    GalleryThumbnailSize
        Defines the width and height of the thumbnail images.

        Defaults to '100x75'

    GalleryThumbnailSizeLS
        If set to '1', GalleryThumbnailSize is the long and the short side
        of the thumbnail image instead of the width and height.

        Defaults to '0'.

    GalleryCopyrightImage
        Image you want to blend into your images in the lower right corner.
        This could be a transparent png saying "copyright my name 2001".

        Optional.

    GalleryWrapNavigation
        Make the navigation in the picture view wrap around (So Next at the
        end displays the first picture, etc.)

        Set to 1 or 0, default is 0

    GalleryAllowOriginal
        Allow the user to download the Original picture without resizing or
        putting the CopyrightImage on it.

        Set to 1 or 0, default is 0

    GallerySlideshowIntervals
        With this option you can configure which intervals can be selected
        for a slideshow. The default is '3 5 10 15 30'

    GallerySortBy
        Instead of the default filename ordering you can sort by any stat
        attribute. For example size, atime, mtime, ctime.

    GalleryMemoize
        Cache EXIF data using Memoize - this will make Apache::Gallery
        faster when many people access the same images, but it will also
        cache EXIF data until the current Apache child dies.

    GalleryEXIFMode
        You can choose how Apache::Gallery should display EXIF info from
        your images.

        The default setting is 'namevalue'. This setting will make
        Apache::Gallery print out the names and values of the EXIF values
        you configure with GalleryInfo. The information will be parsed into
        $INFO in pictureinfo.tpl.

        You can also set it to 'values' which will make A::G parse the
        configured values into the var $EXIFVALUES as 'value | value |
        value'

        If you set this option to 'variables' the items you configure in
        GalleryInfo will be available to your templates as $EXIF_<KEYNAME>
        (in all uppercase). That means that with the default setting
        "Picture Taken => DateTimeOriginal, Flash => Flash" you will have
        the variables $EXIF_DATETIMEORIGINAL and $EXIF_FLASH available to
        your templates. You can place them anywhere you want.

    GalleryRootPath
        Change the location of gallery root. The default is ""

    GalleryRootText
        Change the name that appears as the root element in the menu. The
        default is "root:"

    GalleryMaxThumbnailsPerPage
        This options controls how many thumbnails should be displayed in a
        page. It requires $BROWSELINKS to be in the index.tpl template file.

    GalleryImgFile
        Pattern matching the files you want Apache::Gallery to view in the
        index as thumbnails.

        The default is '\.(jpe?g|png|tiff?|ppm)$'

    GalleryVidFile
        Pattern matching the files you want Apache::Gallery to view in the
        index as videos.

	Only types supported by HTML 5 should be included.

        The default is '\.(ogv|ogx|m4v|mp4|webm)$'

    GalleryDocFile
        Pattern matching the files you want Apache::Gallery to view in the
        index as normal files. All other filetypes will still be served by
        Apache::Gallery but are not visible in the index.

        The default is
        '\.(mp3|m4a|ogg|wav|flac|od.|xlsx?|pptx?|docx?|pdf|rtf|txt|html?|csv|eps)$'

    GalleryTTFDir
        To use the GalleryCopyrightText feature you must set this option to
        the directory where your True Type fonts are stored. No default is
        set.

        Example:

                PerlSetVar      GalleryTTFDir '/usr/share/fonts/'

    GalleryTTFFile
        To use the GalleryCopyrightText feature this option must be set to
        the name of the True Type font you wish to use. Example:

                PerlSetVar      GalleryTTFFile 'verdanab.ttf'

    GalleryTTFSize
        Configure the size of the CopyrightText that will be inserted as
        copyright notice in the corner of your pictures.

        Example:

                PerlSetVar      GalleryTTFSize '10'

    GalleryCopyrightText
        The text that will be inserted as copyright notice.

        Example:

                PerlSetVar      GalleryCopyrightText '(c) Michael Legart'

    GalleryCopyrightColor
        The text color of your copyright notice.

        Examples:

        White: PerlSetVar GalleryCopyrightColor '255,255,255,255'

        Black: PerlSetVar GalleryCopyrightColor '0,0,0,255'

        Red: PerlSetVar GalleryCopyrightColor '255,0,0,255'

        Green: PerlSetVar GalleryCopyrightColor '0,255,0,255'

        Blue: PerlSetVar GalleryCopyrightColor '0,0,255,255'

        Transparent orange: PerlSetVar GalleryCopyrightColor '255,127,0,127'

    GalleryCopyrightBackgroundColor
        The background-color of a GalleryCopyrightText

        r,g,b,a - for examples, see GalleryCopyrightColor

    GalleryQuality
        The quality (1-100) of scaled images

        This setting affects the quality of the scaled images. Set this to a
        low number to reduce the size of the scaled images. Remember to
        clear out your cache if you change this setting. Quality seems to
        default to 75, at least in the jpeg and png loader code in Imlib2
        1.1.0.

        Examples:

        Quality at 50: PerlSetVar GalleryQuality '50'

    GalleryUnderscoresToSpaces
        Set this option to 1 to convert underscores to spaces in the listing
        of directory and file names, as well as in the alt attribute for
        HTML <img> tags.

    GalleryCommentExifKey
        Set this option to e.g. ImageDescription to use this field as
        comments for images.

    GalleryEnableMediaRss
        Set this option to 1 to enable generation of a media RSS feed. This
        can be used e.g. together with the PicLens plugin from
        http://piclens.com

FEATURES
    Ignore directories/files
        To ignore a directory or a file (of any kind, not only images) you
        create a <directory|file>.ignore file.

        To hide a directory or a file from finding it by browsing, whilst
        still allowing direct access if the name is known, create a
        <directory|file>.noindex file.

    Comments
        To include comments for a directory you create a <directory>.comment
        file where the first line can contain "TITLE: New title" which will
        be the title of the page, and a comment on the following lines. To
        include comments for each picture you create files called
        picture.jpg.comment where the first line can contain "TITLE: New
        title" which will be the title of the page, and a comment on the
        following lines.

        Example:

                TITLE: This is the new title of the page
                And this is the comment.<br />
                And this is line two of the comment.

        The visible name of the folder is by default identical to the name
        of the folder, but can be changed by creating a file
        <directory>.folder with the visible name of the folder.

        It is also possible to set GalleryCommentExifKey to the name of an
        EXIF field containing the comment, e.g. ImageDescription. The EXIF
        comment is overridden by the .comment file if it exists.

DEPENDENCIES
    Perl 5
    Apache with mod_perl
    URI::Escape
    Image::Info
    Image::Size
    Text::Template
    Image::Imlib2
    Imlib2 Remember the -dev package when using rpm, deb or other package
    formats!

AUTHOR
    Michael Legart <michael@legart.dk>

    Modifications by Matthew Blissett <matt@blissett.me.uk>.  See source
    code on GitHub for details: <https://github.com/MattBlissett/Apache-Gallery>

COPYRIGHT AND LICENSE
    Copyright © 2001-2011 Michael Legart <michael@legart.dk>

    'default' template designed by Thomas Kjaer <tk@lnx.dk>
    'modern' template designed by Matthew Blissett.

    Apache::Gallery is free software and is released under the Artistic
    License. See http://www.perl.com/language/misc/Artistic.html for
    details.

    The video icons are from the GNOME project. http://www.gnome.org/

THANKS
    Thanks to Thomas Kjaer for templates and design of
    http://apachegallery.dk Thanks to Thomas Eibner and other for patches.
    (See the Changes file)

SEE ALSO
    perl, mod_perl, Image::Imlib2, CGI::FastTemplate, Image::Info, and
    Image::Size.



Changes
Revision history for Perl extension Apache::Gallery.

Not yet released
	- Improved performance by caching HTML
	- Improved performance by skipping checks on images
	- Still to do: make this optional with a config flag
	- Refactoring work for file return and cache directory creation
	- *Temporarily* added timing logging
	- Remove http: from external resources to allow https access
	- Added RFDa properties to 'modern' theme
		- Made 'modern' theme have (almost) valid XHTML+RDFa
	- Extracted bulk of the main handle routine into seperate methods
	- Removed '.cache' path:
		- Old URLs are redirected to new locations where possible
		- /dir/file.jpg is now the image
		- /dir/file.jpg?w=640&h=480 is a scaled image
		- /dir/file is the HTML page
		- /dir/file?width=800 HTML page with chosen width
		- /dir/.points.xml is the points file
		- /dir/.bg-100.jpg is the directory icon
	- Tidied up whitespace and odd bits of style
	- Fixed some bugs
	- OpenGraph metadata <http://ogp.me/>
	- Default 'modern' stylesheet now fills the page with the image
	- Pre-loading of next image (if browser supports it)
	- Quote CSS url(...) references in case filename contains brackets
	- Added altitude to modern template
	- Added Javascript-controlled next/prev photo size to modern template
	- Added support to hide files from browsing, but still allow direct
	  access, with <directory|file>.noindex.
	- Remove mod_perl 1 support
	- Remove old 'bright' and 'new' templates
	- Remove selection mode
	- Default 'modern' stylesheet is now responsive and shows
	  larger map on top or right of screen
	- Show GPS tracks on the map

1.1 Sun Aug  5 23:48:30 BST 2012

	- Added HTML 5 video support
	- Added maps, using EXIF geographical data
	- New "modern" theme, using cover images and maps
	- Minor updates to existing themes
	- Produce album covers (thumbnails using four images from the
	  album) (Matt Blissett)
	- Moved code to Github <https://github.com/MattBlissett/Apache-Gallery>
	  (Matt Blissett)

1.0.2 Wed Jun  8 20:47:16 CEST 2011

	- Extended GalleryUnderscoresToSpaces to filenames as well.
	  (Debian bug #348724, Francesco Potortì)
	- Added txt to GalleryDocFile, the code in Gallery.pm already
	  allowed it.  (Luca Capello)
	- Added text-html.png and text-txt.png icons.
	  (Debian bug #423004, Luca Capello)
	- Fixed counter typo for <directory>.folder in next directory
	  menu item.  (Luca Capello)
	- Added support to ignore items through <directory|file>.ignore.
 	  (Debian bug #619625, Luca Capello)
	- Fixed spelling typos in Gallery.pm and README.  (Luca Capello)
	- GalleryCacheDir defaults to /var/cache/www/ per the FHS-2.3.
	  (Debian bug #337197, Luca Capello)
	- Fixed two minor POD errors (Michael Legart)

1.0.1 Wed Feb 23 20:45:38 CET 2011

	- Added missing template files to MANIFEST

1.0 Tue Feb 22 21:54:31 CET 2011

	- Handle files that match both GalleryDocFile and GalleryImgFile
	  correctly. (Claus Faerber)
	- Only respond to HEAD and GET requests, enabling users to use
	  WebDAV for upload (Andreas Plesner)
	- Added new option GalleryCommentExifKey to get comments from
	  EXIF data (Michael Legart)
	- Added new option GalleryEnableMediaRss to enable generation of
	  a media RSS feed for each directory listing. This works with
	  e.g. the plugin from http://piclens.com to enable 3D viewing
	  of your gallery. (Michael Legart)
	- Make browser-caching work with mod_perl 2. Supports If-None-Match
	  and If-Modified-Since headers. Sets Last-Modified-Date and
	  ETag headers. (Michael Legart)

1.0RC3 Fri Sep 16 10:27:48 CEST 2005

	- Add watermark even when picture doesn't need to be rescaled
	  (Andreas Plesner)
	- Fix logging to work in Apache 1.3 (Andreas Plesner)
	- Bugfix: If only one GallerySize was specified, the image's
	  max width was autmatically added to GallerySizes
	  (Andreas Plesner)
	- Bugfix: Locate thm files if they are called .thm or .THM
		(Michael Legart)

1.0RC2 Wed Jun  1 09:11:50 CEST 2005

	- Added access keys for navigation (Michael Knudsen)

1.0RC1 Tue May 24 13:31:50 CEST 2005

	- Added submit button to form in selection mode (Vlad Marchenko)
	- Added new option GalleryRootPath for use when the gallery
	  is not running from the root of the virtual host (Lubomir Host)
	- Report proper errors when there are problems with templates (Don Armstrong)
	- Support newest mod_perl2 version (Philip Paeps)

0.9.1 Sat Sep 11 23:52:16 CEST 2004

	- Fix mod_perl 1 support that was broken with 0.9 (Michael Legart)

0.9 Sat Sep 11 22:03:20 CEST 2004

	- Handle .thm for all filetypes and not just tiff, gif and png
		(Michael Legart)
	- Make GallerySortBy work for directories (Michael Legart)
	- Make GalleryDirSortBy override GallerySortBy for directories
	  (Andreas Plesner)
	- Report 404 for HEAD requests to non-existing files or directories
	  (Michael Legart)
	- Scale images when they are requested from the cache instead of
	  before the index is displayed. (Michael Legart)
	- Sort pictures the same way when viewing images as when viewing
	  directories (Andreas Plesner)
	- Make the "back" link on the error page work in IE (Michael Legart)
	- Fixed the TITLE tag for folder comments (Ondra Kudlik)
	- New option GalleryUnderscoresToSpaces to convert underscores to
	  spaces in the listing of directories. (Ondra Kudlik)
	- Bugfix for when running outside a virtual host (Jeffrey Hartmann)
	- Support Apache2 on FreeBSD (Jesper Dalberg)
	- Avoid false 404 errors when running under mod_perl 2 (Tom Brown)

0.8 Sun Mar  7 11:22:00 CET 2004

	- Remember choosen width by setting a cookie (Rene Joergensen)
	- Fixed a bug where $EXIFVALUES was left blank if a picture
	  had a comment. (Michael Legart)
	- Fixed a bug where one line comments was showed as $COMMENT
	  (Jesper Skriver)
	- Fixed a bug where comments did not get shown when using
	  the variables exif mode (Thomas L. Kjeldsen)
	- Added two new configuration options GalleryDocFile and
	  GalleryImgFile that makes it possible to configured which

htdocs/modern/modern.js
var map, template, photos, select, tracks;

var mapInitialised = false;
var mapHidden = true;
var infoHidden = true;

OpenLayers.ImgPath = "/ApacheGallery/";

// Initialise index page map.
function initialiseBigMap() {
	map = new OpenLayers.Map();
	map.addControl(new OpenLayers.Control.LayerSwitcher());

	// Load base maps
	var base = new OpenLayers.Layer.OSM(
		"OpenStreetMap",
		// Official OSM tileset as protocol-independent URLs
		[
			'//a.tile.openstreetmap.org/${z}/${x}/${y}.png',
			'//b.tile.openstreetmap.org/${z}/${x}/${y}.png',
			'//c.tile.openstreetmap.org/${z}/${x}/${y}.png'
		],
		null);
	var ghyb = new OpenLayers.Layer.Google(
		"Google Hybrid",
		{type: google.maps.MapTypeId.HYBRID, numZoomLevels: 20}
	);
	var gsat = new OpenLayers.Layer.Google(
		"Google Satellite",
		{type: google.maps.MapTypeId.SATELLITE, numZoomLevels: 22}
	);

	// Define marker style
	var style = new OpenLayers.Style({
		pointRadius: "${radius}",
		fillColor: "#ff00ff",
		fillOpacity: 0.8,
		strokeColor: "#000000",
		strokeWidth: 2,
		strokeOpacity: 0.8
	}, {
		context: {
			radius: function(feature) {
				return Math.min(feature.attributes.count, 7) + 3;
			}
		}
	});

	// Load photos
	photos = new OpenLayers.Layer.Vector("Photos", {
		strategies: [
			new OpenLayers.Strategy.Fixed(),
			new OpenLayers.Strategy.Cluster()
		],
		protocol: new OpenLayers.Protocol.HTTP({
			url: ".points.xml",
			format: new OpenLayers.Format.GML()
		}),
		styleMap: new OpenLayers.StyleMap({
			"default": style,
			"select": {
				fillColor: "#00ffff",
				strokeColor: "#0000ff"
			}
		})
	});
	photos.events.register('loadend', photos, photosLoaded);

	// Loads tracks
	trackColours = ['purple', 'cyan', 'magenta'];
	tracks = [];
	for (i = 0; i < availableTracks.length; i++) {
		tracks[i] = new OpenLayers.Layer.Vector(availableTracks[i], {
			strategies: [new OpenLayers.Strategy.Fixed()],
			protocol: new OpenLayers.Protocol.HTTP({
				url: availableTracks[i],
				format: new OpenLayers.Format.GPX()
			}),
			style: {
				strokeColor: trackColours[i%trackColours.length],
				strokeWidth: 5,
				strokeOpacity: 0.75
			},
			projection: new OpenLayers.Projection("EPSG:4326")
		});
		console.log(i+"th track: "+availableTracks[i]);
		console.log(tracks[i]);
	}

	// Add layers to map
	map.addLayers([base, photos, ghyb, gsat]);
	map.addLayers(tracks);

	// Define select (hover) on markers behaviour
	select = new OpenLayers.Control.SelectFeature(
		photos, {hover: true}
	);
	map.addControl(select);
	select.activate();
	photos.events.on({"featureselected": highlight});
}

// Undims all photos
function undimAll() {
	$('#files a').removeClass("dimmed");
	for (i = 0; i < photos.features.length; i++) {
		var feature = photos.features[i];
		select.unselect(feature);
	}
}

// Called when photos have loaded (seems to be called after a single photo is loaded)
function photosLoaded(event) {
	if (photos.features.length > 0) {
		console.log("At least "+photos.features.length+" photos");
		initMapElements();

		// No idea why it's necessary to call this twice.  Calling once centres on the first cluster,
		// called a second time all the other clusters are shown too.
		map.zoomToExtent(event.object.getDataExtent());
		map.zoomToExtent(event.object.getDataExtent());
	}
	else {
		console.log("No photos");
	}
}

// Called by the map if there are geotagged photos
function initMapElements() {
	// Put the map in its div
	$('#map').css("display","block");
	map.render("map");

	// Set up events to highlight cluster contaning photos
	$('#files a').bind('mouseover', function() {
		var find = $(this).attr("id");
		//console.log("finding "+find);

		// Select the cluster containing the photograph
		for (i = 0; i < photos.features.length; i++) {
			var feature = photos.features[i];
			select.unselect(feature);
			for (j = 0; j < feature.cluster.length; j++) {
				var cluster = feature.cluster[j];
				//console.log("fid: "+cluster.fid);
				if (cluster.fid == find) {
					select.select(feature);
				}
			}
		}
	});
	// And to un-unhighlight
	$('#map').bind('mouseleave', function(event) {
		undimAll();
	});
	$('#files').bind('mouseleave', function(event) {
		undimAll();
	});

	// Set up events to zoom in to hovered thumbnails
	//$('#files a').bind('mouseover', function() {
	//	var find = $(this).attr("id");
	//});

	$('#map').css("width",$('#mapcontainer').innerWidth());
	map.updateSize();
}

// Creates a button to allow the user to show the large map.
// (First function called)
function createToggleMapButton() {

	$('#map').css("display","none");
	$('#mapcontainer').css("display","block");
	$('#mapcontainer').css("visibility","visible");

	// Set up the 'toggle map' button
	var hide = $('#menu').append('<div id="menuButtons"><a id="toggleMap">&#x1f30d;</a></div>');
	$('#toggleMap').bind('click', toggleMap);
}

// Toggle the display of the map
function toggleMap() {
	// Map is shown on top or right depending on screen size
	var mapClass;
	if ($(window).width() < $(window).height()) {
		mapClass = 'mapOnTop';
		$('#map').css("height",$(window).height()/3.0);
	}
	else {
		mapClass = 'mapOnRight';
		$('#map').css("height",$(window).height());
	}

	if (!mapInitialised) {
		mapHidden = false;
		initialiseBigMap();
		$('#directory').addClass(mapClass);
	}

	if (mapInitialised) {
		mapHidden = !mapHidden;
		if (!mapHidden) {
			$('#map').css("display","block");
			$('#directory').addClass(mapClass);
		}
		else {
			$('#map').css("display","none");
			$('#directory').removeClass(mapClass);
		}
		undimAll();
	}

	tileNicely();

	mapInitialised = true;
}

// Highlights all photos in the cluster.
function highlight(event) {
	// Dim all photos
	$('#files a').addClass("2bdimmed");
	// Un-highlight all photos
	//$('#files a').fadeTo('fast', 0.333);

	// Undim ones in the cluster
	visible = false;
	for (var i = 0; i < event.feature.cluster.length; i++) {
		var fileId = event.feature.cluster[i].attributes.file;
		$(jq(fileId)).removeClass("2bdimmed");
		//$(jq(fileId)).fadeTo('fast', 1);
		visible = visible | isScrolledIntoView($(jq(fileId)));
	}

	// If none of the photos in the cluster are visible scroll to the first.
	var idstring = String(event.feature.cluster[0].attributes.file);
	if (!visible) {
		$('html, body').animate({
			scrollTop: $(jq(idstring)).offset().top - $('#menu').height() - 20
		}, 1000);
	}

	$('#files a.2bdimmed').addClass("dimmed");
	$('#files a:not(a.2bdimmed)').removeClass("dimmed");
	$('#files a').removeClass("2bdimmed");
}

// True if elem is visible in the browser window
function isScrolledIntoView(elem) {
	var scrTop = $(window).scrollTop();
	var scrBottom = scrTop + $(window).height();

	var elemTop = $(elem).offset().top;
	var elemBottom = elemTop + $(elem).height();

	return ((elemTop <= scrBottom) && (elemBottom >= scrTop));
}

// Display a small map on the photo page's info area.
// Initialise the map, and display it.
function smallmap(llat, llong, status) {
	map = new OpenLayers.Map('map', { controls: [] });
	map.addControl(new OpenLayers.Control.MouseToolbar());

	$('#map').width($('#info').width());
	$('#map').height(300);

	var base = new OpenLayers.Layer.OSM();
	map.addLayer(base);

	var lonlat = new OpenLayers.LonLat(llong, llat).transform(
		new OpenLayers.Projection("EPSG:4326"),
		map.getProjectionObject()
	)

	map.setCenter(lonlat, 15);

	// Define mark style
	var style = new OpenLayers.Style({
		pointRadius: "4",
		fillColor: "#ff00ff",
		fillOpacity: 0.8,
		strokeColor: "#000000",
		strokeWidth: 2,
		strokeOpacity: 0.8
	});

	// Add mark to map
	mark = new OpenLayers.Layer.Vector("Photo", {
		styleMap: new OpenLayers.StyleMap({ "default": style })
	});

	var point = new OpenLayers.Geometry.Point(llong, llat);
	var pointFeature = new OpenLayers.Feature.Vector(point.transform(
		new OpenLayers.Projection("EPSG:4326"),
		map.getProjectionObject()
	));

	mark.addFeatures(pointFeature);

	map.addLayer(mark);
}

// Escapes an id containing special characters (e.g. 001.jpg -> #001\\.jpg)
function jq(myid) {
	return '#' + myid.replace(/(:|\.)/g,'\\$1');
}

// Adjust the width parameter of the 'next' and 'previous' links according to the screen size.
function adjustPhotoWidths() {
	if (availablePhotoWidths.length == 0 || (document.getElementById('next') == null && document.getElementById('prev') == null)) {
		return;
	}

	var h = $(window).height();
	var w = $(window).width();

	// Choose the largest width that's not larger than the screen.
	var picw = availablePhotoWidths[availablePhotoWidths.length-1];

	for (var i = availablePhotoWidths.length - 1; i >= 0; i--) {
		if (w > availablePhotoWidths[i]) {
			break;
		}
		picw = availablePhotoWidths[i];
	}

	if (document.getElementById('next') != null) {
		var oldNextLink = $('#next').prop('href');
		var newNextLink = oldNextLink.replace(/width=\d+/, 'width='+picw);
		$('#next').prop('href', newNextLink);
		if (document.getElementById('refresh') != null) {
			var oldRefreshLink = $('#refresh').prop('content');
			var newRefreshLink = oldRefreshLink.replace(/width=\d+/, 'width='+picw);
			$('#refresh').prop('content', newRefreshLink);
		}
	}

	if (document.getElementById('prev') != null) {
		var oldPrevLink = $('#prev').prop('href');
		var newPrevLink = oldPrevLink.replace(/width=\d+/, 'width='+picw);
		$('#prev').prop('href', newPrevLink);
	}

	$("#size").hide();
}

function extractNumericAttribute(elem, attribute) {
	var w = parseInt($(elem).attr(attribute));
	if (w > 0) {
		return w;
	}
	else {
		return 80;
	}
}

// Adjust photo widths and heights to tile nicely across the screen.
function tileNicely() {
	if (document.getElementById('files') == null) {
		return;
	}

	// Target width and height
	var h_s = $(window).height() - 40;
	var w_s = $("#files").width() - 40;

	// Target number of rows shown
	var rows = 2.8;

	// Aim for rows rows of photos of width w_aim and height h_aim
	// h_aim = w_aim/1.5 (assumed ratio of landscape photo)
	// h_s = rows * h_aim = rows * w_aim/1.5
	var h_aim = h_s / rows;
	var w_aim = h_s * 1.5 / rows;

	// How many photos fit in the screen width?
	var n = Math.round(w_s / w_aim);
	if (n < 2) n = 2;

	// Minimum total width, allowing for some scaling up, is thus
	var wt_min = n * w_aim;

	// Load photo sizes (supplied in HTML)
	var p_photos = $("#files a");
	var w_photos = $("#files a").map( function() {
		return extractNumericAttribute(this, 'data-width');
	}).get();
	var h_photos = $("#files a").map( function() {
		return extractNumericAttribute(this, 'data-height');
	}).get();

	// Next row
	var w_row = [];
	var h_row = [];

	var i = 0;
	var j = 0;

	var p = 0;
	while (i < w_photos.length) {
		// So pick photos until ∑w > wt_min.
		w_row[w_row.length] = w_photos[i];
		h_row[h_row.length] = h_photos[i];

		// Sum widths in current row
		w_row_total = 0;
		for (j = 0; j < w_row.length; j++) {
			w_row_total += w_row[j] * h_aim / h_row[j];
		}

		var lastRow = (i + 1 == w_photos.length);

		// We have a row to process
		if (w_row_total > wt_min || lastRow) {
			// Scale the photos
			for (j = 0; j < w_row.length; j++) {
				var factor = h_aim / h_row[j];
				h_row[j] = h_aim;
				w_row[j] = w_row[j] * factor;
			}

			w_row_total = 0;
			for (j = 0; j < w_row.length; j++) {
				w_row_total += w_row[j];
			}

			if (lastRow && w_row_total / w_s < 0.66) {
				w_row_total = w_s;
			}

			// Scale to screen width
			var w_factor = (w_s - (10 * (w_row.length - 1))) / w_row_total;
			for (j = 0; j < w_row.length; j++) {
				h_row[j] = h_row[j] * w_factor;
				w_row[j] = w_row[j] * w_factor;

				var elem = $(p_photos.get(p));

				// Set dimensions
				elem.width(w_row[j]);
				elem.height(h_row[j]);

				// Calculate change to dimensions in provided image URL
				var img_url = elem.attr("data-img");

				// Round to nearest 100 to reduce number of sizes generated by the server
				var new_width = Math.ceil(w_row[j]/100)*100;
				var new_img_url;
				if (img_url.match(/.bg-\d+/)) {
					new_img_url = img_url.replace(/.bg-\d+/, '.bg-'+new_width);
					resized = true;
				}
				else if (img_url.match(/w=\d+/)) {
					new_img_url = img_url.replace(/w=\d+/, 'w='+new_width).replace(/h=\d+/, 'h='+Math.round(h_row[j]*new_width/w_row[j]));
					resized = true;
				}

				if (resized) {
					// Set style
					var old_style = elem.attr('style');
					var new_style = old_style + "; background-image: url('" + new_img_url + "'), url('/ApacheGallery/modern/squares.svg');";
					elem.attr('style', new_style);

					// Event to remove SVG spinner once image is loaded, as it uses a lot of CPU even when obscured.
					$('<img/>').attr('src', new_img_url).data('target', elem).load(function() {
						$(this).data('target').css('background-image', "url('"+$(this).attr('src')+"')");
						$(this).remove();
					});
				}

				// Change CSS rule since the sizes are different.
				var elemSpan = elem.children('span').first();
				elemSpan.css("margin", "0");
				elemSpan.css("transform", "translateX("+(w_row[j]/2)+"px) translateX(-50%) translateY("+(h_row[j]/2)+"px) translateY(-50%) rotate("+(Math.random()*0.1-0.05)+"rad)");

				p++;
			}

			// Start a new row
			w_row = [];
			h_row = [];
		}
		i++;
	}
}

// Creates a button to allow the user to toggle the photo information.
// (First function called)
function createToggleInfoButton() {
	// Set up the 'toggle info' button
	var hide = $('#menuButtons').append('<ul id="toggleInfo"><li>&#x1f4f7;</li></ul>');
	$('#toggleInfo').bind('click', toggleInfo);
}

// Toggle the display of the photo information
function toggleInfo() {
	if (document.getElementById('info') == null) {
		return;
	}

	// Info is shown on the right
	var infoClass;
	infoClass = 'infoOnRight';

	infoHidden = !infoHidden;
	if (!infoHidden) {
		$('body').addClass(infoClass);
		smallmap(llat, llong, status);
	}
	else {
		$('body').removeClass(infoClass);
	}

	infoInitialised = true;
	put('infoShown', !infoHidden);
}

var get = function (key) {
	return window.localStorage ? window.localStorage[key] : null;
}

var put = function (key, value) {
	if (window.localStorage) {
		window.localStorage[key] = value;
	}
}

$(window).resize(function() {
	adjustPhotoWidths();
	tileNicely();
	if (!mapHidden) {
		$('#map').css("width",$('#mapcontainer').innerWidth());
		map.updateSize();
	}
});

$(document).ready(function() {
	adjustPhotoWidths();
	tileNicely();

	if (get('infoShown') == 'true') {
		toggleInfo();
	}

	if (get('mapShown') == 'true') {
		toggleInfo();
	}
});
